actors/miner/src/lib.rs

```diff
@@ -3273,36 +3286,188 @@ impl Actor {
+                    pending_beneficiary_term.approved_by_beneficiary = true;
+                }
+                info.pending_beneficiary_term = Some(pending_beneficiary_term);
+            } else if let Some(pending_term) = &info.pending_beneficiary_term {
+                if pending_term.new_beneficiary != new_beneficiary {
+                    return Err(actor_error!(
+                        illegal_argument,
+                        "new beneficiary address must be equal expect {}, but got {}",
+                        pending_term.new_beneficiary,
+                        params.new_beneficiary
+                    ));
+                }
+                if pending_term.new_quota != params.new_quota {
+                    return Err(actor_error!(
+                        illegal_argument,
+                        "new beneficiary quota must be equal expect {}, but got {}",
+                        pending_term.new_quota,
+                        params.new_quota
+                    ));
+                }
+                if pending_term.new_expiration != params.new_expiration {
+                    return Err(actor_error!(
+                        illegal_argument,
+                        "new beneficiary expire date must be equal expect {}, but got {}",
+                        pending_term.new_expiration,
+                        params.new_expiration
+                    ));
+                }
+            } else {
+                return Err(actor_error!(forbidden, "No changeBeneficiary proposal exists"));
+            }
+
+            if let Some(pending_term) = info.pending_beneficiary_term.as_mut() {
+                if rt.message().caller() == info.beneficiary {
+                    pending_term.approved_by_beneficiary = true
+                }
+
+                if rt.message().caller() == new_beneficiary {
+                    pending_term.approved_by_nominee = true
+                }
+
+                if pending_term.approved_by_beneficiary && pending_term.approved_by_nominee {
+                    //approved by both beneficiary and nominee
+                    if new_beneficiary != info.beneficiary {
+                        //if beneficiary changes, reset used_quota to zero
+                        info.beneficiary_term.used_quota = TokenAmount::zero();
+                    }
+                    info.beneficiary = new_beneficiary;
+                    info.beneficiary_term.quota = pending_term.new_quota.clone();
+                    info.beneficiary_term.expiration = pending_term.new_expiration;
+                    // clear the pending proposal
+                    info.pending_beneficiary_term = None;
+                }
+            }
+
+            state.save_info(rt.store(), &info).map_err(|e| {
+                e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "failed to save miner info")
+            })?;
+            Ok(())
```

Comment:
I haven't followed the discussion on the FIP, was the behaviour when the `change_beneficiary` is called by 3rd party (neither the miner nor the new beneficiary) raised? Someone being a 3rd party can get an `Ok` result code from this call.
While harmless it could lead to confusion, similar to the thing where you could send yourself virtual billions of Filecoin (you didn't get richer but it looked funky).

Reply:
Agree we should fix this and return `USR_FORBIDDEN`

Reply:
done,

Reply:
Sorry, I didn't read the code enough before responding above. I don't think @Kubuxu 's comment was in relation to the lines immediately above, but the whole method. This `state.save_info` call should go back to using USR_ILLEGAL_STATE, because it won't catch the case in question.

If there is a pending change, but the caller is neither new nor old beneficiary, then all the checks pass but the state update is a no-op. I think around line 3390 we need an explicit check that the caller is either the old or new beneficiary for the proposal that is loaded from state, and FORBIDDEN if not.

Reply:
https://github.com/filecoin-project/builtin-actors/pull/496/commits/50a0ce4cc92ecc9848b6a9a4e3b99af4c995b7f8

Reply:
done